src/irmin-pack/unix/file_manager.ml

```diff
@@ -749,4 +760,44 @@ struct
     | T15 ->
         (* Unreachable *)
         assert false
+
+  (* TODO : this does not work yet with GC. *)
+  let split t : (unit, [> Errs.t ]) result =
+    let open Result_syntax in
+    (* Step 1. Fsync last chunk *)
+    let* () = Suffix.fsync t.suffix in
+
+    (* Step 2. Create a new chunk file *)
+    let auto_flush_threshold =
+      match Suffix.auto_flush_threshold t.suffix with
+      | None -> assert false
+      | Some x -> x
+    in
+    let cb _ = suffix_requires_a_flush_exn t in
+    let* () =
+      Suffix.add_chunk ~auto_flush_threshold
+        ~auto_flush_procedure:(`External cb) t.suffix
+    in
+
+    (* Step 3. Update the control file *)
+    let* () =
+      let pl = Control.payload t.control in
+      let pl =
+        let open Payload in
+        let chunk_num = pl.chunk_num + 1 in
+        let suffix_end_poff = Int63.zero in
+        { pl with chunk_num; suffix_end_poff }
+      in
+      [%log.debug "split: writing new control_file chunk_num %d" pl.chunk_num];
+      Control.set_payload t.control pl
+    in
+    (* Step 4. Fsync the control file *)
+    Control.fsync t.control
```

Comment:
I'm still wondering about this `fsync`. We update the control file in other places without `fsync`ing (for instance, in `swap`) -- should we break that here?

In the event that the control file data does not make it to disk, the following things would happen:

1. There would be a dangling chunk that is removed on next open by `cleanup`
2. A future GC would be non-optimal and a chunk would live an extra GC cycle before being deleted.

Point 1 is okay (I think) and point 2 is less great since it means more disk space usage. I don't have a strong opinion here but maybe we leave it and see what impact it has on benchmarks (?).

Reply:
Yeah, I'm not sure about the `fsync` calls here neither
- `fsync` on one file might lead to `fsync` for all files on some filesystems (see https://github.com/mirage/irmin/issues/1959#issuecomment-1180569176);
- we might have to call `fsync`on the parent directory because, according to https://man7.org/linux/man-pages/man2/fdatasync.2.html:  
```
       Calling fsync() does not necessarily ensure that the entry in the
       directory containing the file has also reached disk.  For that an
       explicit fsync() on a file descriptor for the directory is also
       needed. 
``` 
So I think we should remove the `fsyncs` from the main process, and do an `fsync` on the root at the end of each Gc (similar to what we have now). We only guarantee that data is not lost if it was added before the last gc.

Reply:
Are you also saying that you want to remove the `fsync` on the suffix before creating the new chunk? I still think that is a good idea, but either way, I think we have identified a place that needs more careful consideration.

Also, you might be saying this for `fsync`ing the directory but just to be on the same page, I think that only applies when you are adding new files since the language of the manual is `ensure that the entry in the directory`. From [Ensuring data reaches disk](https://lwn.net/Articles/457667/):

> The more subtle usages deal with newly created files, or overwriting existing files. A newly created file may require an fsync() of not just the file itself, but also of the directory in which it was created (since this is where the file system looks to find your file). This behavior is actually file system (and mount option) dependent. You can either code specifically for each file system and mount option combination, or just perform fsync() calls on the directories to ensure that your code is portable. 

This would apply for the new chunk but not the control file.

Reply:
Indeed, I think we need to think a bit more about `fsync` here. For now I removed it. 

> Are you also saying that you want to remove the `fsync` on the suffix before creating the new chunk? I still think that is a good idea (I haven't thought through all the consequences otherwise), but either way, I think we have identified a place that needs more careful consideration.

Yes, I think it's not a good idea to call `fsync` at all in the main process. If we leave all `fsync` to the gc process, we can guarantee durability for data up to the last gc. I think this is what we have now too?

> Also, you might be saying this for `fsync`ing the directory but just to be on the same page, I think that only applies when you are adding new files since the language of the manual is `ensure that the entry in the directory`.
This would apply for the new chunk but not the control file.

Indeed, it only applies to new chunk files. But our API lets the user call `split` several times before a call to `gc`. If we do add `fsync` to the control file here, we have to call `fsync` for all the directory.

Reply:
:+1: I definitely understand the hesitancy to introduce an `fsync` in the main process. My main concern is that if we do not `fsync` here, we have no control file field to help us after a crash. If the `N-1` chunk didn't flush all the way to disk, it will open fine (will pass the consistency check bc we will just pass the length of the chunk as `end_poff`) but actually not be correct. This has me thinking more about `suffix_end_poff` and whether we should consider moving to storing the length of the suffix that has been persisted in order to check that for consistency on open. Let's keep this as an open discussion (particularly if we need to track more information in the control file).